From 8fb2f4ad90992b0df7849202f13f1cd5d9bbcb5b Mon Sep 17 00:00:00 2001
From: Lars Uebernickel <lars@uebernic.de>
Date: Thu, 27 Feb 2014 13:34:43 +0100
Subject: [PATCH] Restore traditional menu bar

This is probably only useful for Unity, as it relies on unity-gtk-module
exporting the GtkMenu.
---
 src/nautilus-application.c         | 38 +++++++++++-----
 src/nautilus-canvas-view-ui.xml    | 19 ++++++++
 src/nautilus-directory-view-ui.xml | 79 ++++++++++++++++++++++++++++++++
 src/nautilus-list-view-ui.xml      |  7 +++
 src/nautilus-shell-ui.xml          | 92 ++++++++++++++++++++++++++++++++++++++
 src/nautilus-toolbar.c             | 64 +++++++++++++++-----------
 src/nautilus-window-menus.c        | 56 ++++++++++++-----------
 src/nautilus-window-private.h      |  3 +-
 src/nautilus-window.c              | 10 +++--
 9 files changed, 299 insertions(+), 69 deletions(-)

Index: b/src/nautilus-application-actions.c
===================================================================
--- a/src/nautilus-application-actions.c
+++ b/src/nautilus-application-actions.c
@@ -258,8 +258,8 @@
 void
 nautilus_init_application_actions (NautilusApplication *app)
 {
-	GtkBuilder *builder;
-	GError *error = NULL;
+	gboolean show_app_menu;
+	gboolean show_menubar;
 	const gchar *debug_no_app_menu;
 
 	g_action_map_add_action_entries (G_ACTION_MAP (app),
@@ -267,18 +267,31 @@
 					 app);
 	gtk_application_add_accelerator (GTK_APPLICATION (app), "F10", "win.gear-menu", NULL);
 
-	builder = gtk_builder_new ();
-	gtk_builder_add_from_resource (builder, "/org/gnome/nautilus/nautilus-app-menu.ui", &error);
+	g_object_get (gtk_settings_get_default (),
+		      "gtk-shell-shows-app-menu", &show_app_menu,
+		      "gtk-shell-shows-menubar", &show_menubar,
+		      NULL);
 
-	if (error == NULL) {
-		gtk_application_set_app_menu (GTK_APPLICATION (app),
-					      G_MENU_MODEL (gtk_builder_get_object (builder, "app-menu")));
-	} else {
-		g_critical ("Unable to add the application menu: %s\n", error->message);
-		g_error_free (error);
-	}
+	/* Only add the app menu when the shell doesn't show the menu
+	 * bar, because we merge the app menu into the menu bar.
+	 */
+	if (show_app_menu && !show_menubar) {
+		GtkBuilder *builder;
+		GError *error = NULL;
+
+		builder = gtk_builder_new ();
+		gtk_builder_add_from_resource (builder, "/org/gnome/nautilus/nautilus-app-menu.ui", &error);
 
-	g_object_unref (builder);
+		if (error == NULL) {
+			gtk_application_set_app_menu (GTK_APPLICATION (app),
+						      G_MENU_MODEL (gtk_builder_get_object (builder, "app-menu")));
+		} else {
+			g_critical ("Unable to add the application menu: %s\n", error->message);
+			g_error_free (error);
+		}
+
+		g_object_unref (builder);
+	}
 
 	debug_no_app_menu = g_getenv ("NAUTILUS_DEBUG_NO_APP_MENU");
 	if (debug_no_app_menu) {
Index: b/src/nautilus-canvas-view-ui.xml
===================================================================
--- a/src/nautilus-canvas-view-ui.xml
+++ b/src/nautilus-canvas-view-ui.xml
@@ -1,4 +1,23 @@
 <ui>
+  <menubar name="menubar">
+    <menu action="View">
+      <placeholder name="View Items Placeholder">
+        <placeholder action="Arrange Items">
+          <menuitem name="Manual Layout" action="Manual Layout"/>
+          <placeholder name="Auto Layout">
+            <menuitem name="Sort by Name" action="Sort by Name"/>
+            <menuitem name="Sort by Size" action="Sort by Size"/>
+            <menuitem name="Sort by Type" action="Sort by Type"/>
+            <menuitem name="Sort by Modification Date" action="Sort by Modification Date"/>
+            <menuitem name="Sort by Trash Time" action="Sort by Trash Time"/>
+          </placeholder>
+          <separator name="Layout separator"/>
+          <menuitem name="Reversed Order" action="Reversed Order"/>
+        </placeholder>
+        <menuitem name="Keep Aligned" action="Keep Aligned"/>
+      </placeholder>
+    </menu>
+  </menubar>
   <popup name="ViewMenu" accelerators="true">
     <placeholder name="Layout Options">
       <menuitem name="Manual Layout" action="Manual Layout"/>
Index: b/src/nautilus-directory-view-ui.xml
===================================================================
--- a/src/nautilus-directory-view-ui.xml
+++ b/src/nautilus-directory-view-ui.xml
@@ -4,6 +4,85 @@
 <accelerator action="OpenCloseParent"/>
 <accelerator action="PropertiesAccel"/>
 <accelerator action="RenameSelectAll"/>
+<menubar name="menubar">
+  <menu action="File">
+    <placeholder name="New Items Placeholder">
+      <menuitem name="New Folder" action="New Folder"/>
+      <menuitem name="New Folder with Selection" action="New Folder with Selection"/>
+      <menu action="New Documents">
+        <placeholder name="New Documents Placeholder"/>
+        <separator name="After New Documents"/>
+      </menu>
+    </placeholder>
+    <placeholder name="Open Placeholder">
+      <menuitem name="Open" action="Open"/>
+      <menuitem name="OpenInNewTab" action="OpenInNewTab"/>
+      <menuitem name="OpenAlternate" action="OpenAlternate"/>
+      <placeholder name="Applications Placeholder">
+      </placeholder>
+      <menu action="Open With">
+        <placeholder name="Applications Placeholder"/>
+        <separator name="Open With Separator"/>
+        <menuitem name="OtherApplication" action="OtherApplication1"/>
+      </menu>
+      <placeholder name="OtherApplicationPlaceholder">
+        <menuitem name="OtherApplication" action="OtherApplication2"/>
+      </placeholder>
+      <menu action="Scripts">
+        <placeholder name="Scripts Placeholder"/>
+        <separator name="After Scripts"/>
+        <menuitem name="Open Scripts Folder" action="Open Scripts Folder"/>
+      </menu>
+    </placeholder>
+    <placeholder name="File Items Placeholder">
+      <menuitem name="Self Mount Volume" action="Self Mount Volume"/>
+      <menuitem name="Self Unmount Volume" action="Self Unmount Volume"/>
+      <menuitem name="Self Eject Volume" action="Self Eject Volume"/>
+      <menuitem name="Self Start Volume" action="Self Start Volume"/>
+      <menuitem name="Self Stop Volume" action="Self Stop Volume"/>
+      <menuitem name="Self Poll" action="Self Poll"/>
+      <separator name="Properties Separator"/>
+      <menuitem name="Properties" action="Properties"/>
+    </placeholder>
+    <placeholder name="Global File Items Placeholder">
+      <menuitem name="Empty Trash" action="Empty Trash"/>
+      <menuitem name="Save Search" action="Save Search"/>
+      <menuitem name="Save Search As" action="Save Search As"/>
+    </placeholder>
+  </menu>
+  <menu action="Edit">
+    <placeholder name="Clipboard Actions">
+      <menuitem name="Cut" action="Cut"/>
+      <menuitem name="Copy" action="Copy"/>
+      <menuitem name="Paste" action="Paste"/>
+    </placeholder>
+    <placeholder name="Undostack Actions">
+      <menuitem name="Undo" action="Undo"/>
+      <menuitem name="Redo" action="Redo"/>
+    </placeholder>
+    <placeholder name="Select Items">
+      <menuitem name="Select All" action="Select All"/>
+      <menuitem name="Select Pattern" action="Select Pattern"/>
+      <menuitem name="Invert Selection" action="Invert Selection"/>
+    </placeholder>
+    <placeholder name="File Items Placeholder">
+      <menuitem name="Create Link" action="Create Link"/>
+      <menuitem name="Rename" action="Rename"/>
+    </placeholder>
+    <placeholder name="Dangerous File Items Placeholder">
+      <menuitem name="Trash" action="Trash"/>
+      <menuitem name="Delete" action="Delete"/>
+      <menuitem name="Restore From Trash" action="Restore From Trash"/>
+    </placeholder>
+    <placeholder name="Extension Actions"/>
+  </menu>
+  <menu action="View">
+    <placeholder name="View Preferences Placeholder">
+      <menuitem name="Reset to Defaults" action="Reset to Defaults"/>
+      <menuitem name="Show Hidden Files" action="Show Hidden Files"/>
+    </placeholder>
+  </menu>
+</menubar>
 <popup name="ViewMenu" accelerators="true">
   <placeholder name="View Details">
     <menuitem name="Reset to Defaults" action="Reset to Defaults"/>
Index: b/src/nautilus-list-view-ui.xml
===================================================================
--- a/src/nautilus-list-view-ui.xml
+++ b/src/nautilus-list-view-ui.xml
@@ -1,4 +1,11 @@
 <ui>
+  <menubar name="menubar">
+    <menu action="View">
+      <placeholder name="View Items Placeholder">
+        <menuitem name="Visible Columns" action="Visible Columns"/>
+      </placeholder>
+    </menu>
+  </menubar>
   <popup name="ViewMenu" accelerators="true">
     <placeholder name="View Details">
       <menuitem name="Visible Columns" action="Visible Columns"/>
Index: b/src/nautilus-shell-ui.xml
===================================================================
--- a/src/nautilus-shell-ui.xml
+++ b/src/nautilus-shell-ui.xml
@@ -74,4 +74,96 @@
         <separator/>
 	<placeholder name="After Zoom Items"/>
 </popup>
+<menubar name="menubar">
+  <menu action="File">
+    <placeholder name="New Items Placeholder">
+      <menuitem name="New Window" action="New Window"/>
+      <menuitem name="New Tab" action="New Tab"/>
+    </placeholder>
+    <separator/>
+    <placeholder name="Open Placeholder"/>
+    <separator/>
+    <placeholder name="Location Placeholder"/>
+    <menuitem name="Connect to Server" action="Connect to Server"/>
+    <separator/>
+    <placeholder name="File Items Placeholder"/>
+    <separator/>
+    <placeholder name="Global File Items Placeholder"/>
+    <separator/>
+    <placeholder name="Extension Actions"/>
+    <separator/>
+    <menuitem name="Close" action="Close"/>
+    <placeholder name="Close Items Placeholder">
+      <menuitem name="Close All Windows" action="Close All Windows"/>
+    </placeholder>
+  </menu>
+  <menu action="Edit">
+    <placeholder name="Undostack Actions"/>
+    <separator/>
+    <placeholder name="Clipboard Actions">
+    </placeholder>
+    <separator/>
+    <placeholder name="Select Items"/>
+    <separator/>
+    <placeholder name="File Items Placeholder"/>
+    <separator/>
+    <placeholder name="Dangerous File Items Placeholder"/>
+    <separator/>
+    <placeholder name="Edit Items Placeholder"/>
+    <placeholder name="Global Edit Items Placeholder"/>
+    <separator/>
+    <placeholder name="Extension Actions"/>
+    <separator/>
+    <menuitem name="Preferences" action="Preferences"/>
+  </menu>
+  <menu action="View">
+    <menuitem name="Stop" action="Stop"/>
+    <menuitem name="Reload" action="Reload"/>
+    <separator/>
+    <menuitem name="View List" action="View List" />
+    <menuitem name="View Grid" action="View Grid" />
+    <separator />
+    <placeholder name="View Items Placeholder"/>
+    <separator/>
+    <placeholder name="View Preferences Placeholder"/>
+    <separator/>
+    <menuitem name="Show Hide Sidebar" action="Show Hide Sidebar"/>
+    <separator/>
+    <placeholder name="Zoom Items Placeholder">
+      <menuitem name="Zoom In" action="Zoom In"/>
+      <menuitem name="Zoom Out" action="Zoom Out"/>
+      <menuitem name="Zoom Normal" action="Zoom Normal"/>
+    </placeholder>
+    <placeholder name="View Choices">
+      <separator name="Before Short List"/>
+      <placeholder name="Short List"/>
+    </placeholder>
+  </menu>
+  <menu action="Go">
+    <menuitem name="Up" action="Up" />
+    <menuitem name="Back" action="Back" />
+    <menuitem name="Forward" action="Forward" />
+    <separator />
+    <menuitem name="Enter Location" action="Enter Location" />
+    <menuitem name="Find" action="Search" />
+  </menu>
+  <placeholder name="Other Menus">
+    <menu action="Bookmarks">
+      <menuitem name="Add Bookmark" action="Add Bookmark"/>
+      <menuitem name="Edit Bookmark" action="Edit Bookmarks"/>
+      <separator/>
+      <placeholder name="Bookmarks Placeholder"/>
+    </menu>
+
+  </placeholder>
+  <menu action="Help">
+    <menuitem action="NautilusHelp"/>
+    <menuitem action="NautilusHelpSearch"/>
+    <menuitem action="NautilusHelpSort"/>
+    <menuitem action="NautilusHelpLost"/>
+    <menuitem action="NautilusHelpShare"/>
+    <separator/>
+    <menuitem name="About Nautilus" action="About Nautilus"/>
+  </menu>
+</menubar>
 </ui>
Index: b/src/nautilus-toolbar.c
===================================================================
--- a/src/nautilus-toolbar.c
+++ b/src/nautilus-toolbar.c
@@ -377,6 +377,18 @@
         return FALSE;
 }
 
+static gboolean
+nautilus_toolbar_show_menu_buttons ()
+{
+	gboolean show_menubar;
+
+	g_object_get (gtk_settings_get_default (),
+		      "gtk-shell-shows-menubar", &show_menubar,
+		      NULL);
+
+	return !show_menubar;
+}
+
 static void
 nautilus_toolbar_constructed (GObject *obj)
 {
@@ -426,16 +438,18 @@
 	self->priv->location_entry = nautilus_location_entry_new ();
 	gtk_header_bar_pack_start (GTK_HEADER_BAR (toolbar), self->priv->location_entry);
 
-	/* Action Menu */
-	button = toolbar_create_toolbutton (self, TRUE, FALSE, "open-menu-symbolic", _("Location options"));
-	gtk_widget_set_valign (button, GTK_ALIGN_CENTER);
-	menu = gtk_ui_manager_get_widget (ui_manager, "/ActionMenu");
-	gtk_widget_set_halign (menu, GTK_ALIGN_END);
-	gtk_menu_button_set_popup (GTK_MENU_BUTTON (button), menu);
-	gtk_actionable_set_action_name (GTK_ACTIONABLE (button), "win.gear-menu");
-        g_signal_connect (menu, "key-press-event", G_CALLBACK (gear_menu_key_press), self);
+	if (nautilus_toolbar_show_menu_buttons ()) {
+		/* Action Menu */
+		button = toolbar_create_toolbutton (self, TRUE, FALSE, "open-menu-symbolic", _("Location options"));
+		gtk_widget_set_valign (button, GTK_ALIGN_CENTER);
+		menu = gtk_ui_manager_get_widget (ui_manager, "/ActionMenu");
+		gtk_widget_set_halign (menu, GTK_ALIGN_END);
+		gtk_menu_button_set_popup (GTK_MENU_BUTTON (button), menu);
+		gtk_actionable_set_action_name (GTK_ACTIONABLE (button), "win.gear-menu");
+	        g_signal_connect (menu, "key-press-event", G_CALLBACK (gear_menu_key_press), self);
 
-	gtk_header_bar_pack_end (GTK_HEADER_BAR (toolbar), button);
+		gtk_header_bar_pack_end (GTK_HEADER_BAR (toolbar), button);
+	}
 
 	/* View buttons */
 	box = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 0);
@@ -446,11 +460,14 @@
 	button = toolbar_create_toolbutton (self, FALSE, TRUE, NAUTILUS_ACTION_VIEW_GRID, NULL);
 	gtk_widget_set_valign (button, GTK_ALIGN_CENTER);
 	gtk_container_add (GTK_CONTAINER (box), button);
-	button = toolbar_create_toolbutton (self, TRUE, FALSE, "go-down-symbolic", _("View options"));
-	gtk_widget_set_valign (button, GTK_ALIGN_CENTER);
-	gtk_container_add (GTK_CONTAINER (box), button);
-	menu = gtk_ui_manager_get_widget (ui_manager, "/ViewMenu");
-	gtk_menu_button_set_popup (GTK_MENU_BUTTON (button), menu);
+
+	if (nautilus_toolbar_show_menu_buttons ()) {
+		button = toolbar_create_toolbutton (self, TRUE, FALSE, "go-down-symbolic", _("View options"));
+		gtk_widget_set_valign (button, GTK_ALIGN_CENTER);
+		gtk_container_add (GTK_CONTAINER (box), button);
+		menu = gtk_ui_manager_get_widget (ui_manager, "/ViewMenu");
+		gtk_menu_button_set_popup (GTK_MENU_BUTTON (button), menu);
+	}
 
 	gtk_style_context_add_class (gtk_widget_get_style_context (box),
 				     GTK_STYLE_CLASS_RAISED);
Index: b/src/nautilus-window-menus.c
===================================================================
--- a/src/nautilus-window-menus.c
+++ b/src/nautilus-window-menus.c
@@ -473,6 +473,11 @@
 }
 
 static const GtkActionEntry main_entries[] = {
+  /* name, stock id, label */  { "File", NULL, N_("_File") },
+  /* name, stock id, label */  { "Edit", NULL, N_("_Edit") },
+  /* name, stock id, label */  { "View", NULL, N_("_View") },
+  /* name, stock id, label */  { "Go", NULL, N_("Go") },
+  /* name, stock id, label */  { "Bookmarks", NULL, N_("Bookmarks") },
   /* name, stock id, label */  { "Help", NULL, N_("_Help") },
   /* name, stock id */         { NAUTILUS_ACTION_CLOSE, NULL,
   /* label, accelerator */       N_("_Close"), "<control>W",
@@ -628,7 +633,7 @@
 static const GtkRadioActionEntry view_radio_entries[] = {
 	{ NAUTILUS_ACTION_VIEW_LIST, "view-list-symbolic", N_("List"),
 	  "<control>1", N_("View items as a list"), 0 },
-	{ NAUTILUS_ACTION_VIEW_GRID, "view-grid-symbolic", N_("List"),
+	{ NAUTILUS_ACTION_VIEW_GRID, "view-grid-symbolic", N_("Icons"),
 	  "<control>2", N_("View items as a grid of icons"), 1 }
 };
 
@@ -662,14 +667,19 @@
  * 
  * Create and install the set of menus for this window.
  * @window: A recently-created NautilusWindow.
+ *
+ * Returns: (allow-none): the menubar to be installed on the main
+ * window, or %NULL
  */
-void 
+GtkWidget *
 nautilus_window_initialize_menus (NautilusWindow *window)
 {
 	GtkActionGroup *action_group;
 	GtkUIManager *ui_manager;
 	GtkAction *action;
 	gint i;
+	gboolean show_app_menu;
+	gboolean show_menubar;
 
 	window->details->ui_manager = gtk_ui_manager_new ();
 	ui_manager = window->details->ui_manager;
@@ -728,7 +738,17 @@
 			  G_CALLBACK (connect_proxy_cb), window);
 
 	/* add the UI */
+	g_object_get (gtk_settings_get_for_screen (gtk_widget_get_screen (GTK_WIDGET (window))),
+		      "gtk-shell-shows-menubar", &show_menubar,
+		      NULL);
 	gtk_ui_manager_add_ui_from_resource (ui_manager, "/org/gnome/nautilus/nautilus-shell-ui.xml", NULL);
+	/* Only return a menbar if it is exported to the shell.
+	 * Otherwise the view/action menu buttons will be used
+	 */
+	if (show_menubar)
+		return gtk_ui_manager_get_widget (ui_manager, "/menubar");
+	else
+		return NULL;
 }
 
 static GList *
Index: b/src/nautilus-window-private.h
===================================================================
--- a/src/nautilus-window-private.h
+++ b/src/nautilus-window-private.h
@@ -108,7 +108,7 @@
 
 /* window menus */
 void               nautilus_window_initialize_actions                    (NautilusWindow    *window);
-void               nautilus_window_initialize_menus                      (NautilusWindow    *window);
+GtkWidget *        nautilus_window_initialize_menus                      (NautilusWindow    *window);
 
 void               nautilus_window_update_show_hide_menu_items           (NautilusWindow     *window);
 
Index: b/src/nautilus-window.c
===================================================================
--- a/src/nautilus-window.c
+++ b/src/nautilus-window.c
@@ -1439,6 +1439,7 @@
 	GtkWidget *grid;
 	NautilusWindowSlot *slot;
 	NautilusApplication *application;
+	GtkWidget *menubar;
 	const gchar * desktop;
 
 	window = NAUTILUS_WINDOW (self);
@@ -1460,9 +1461,14 @@
 	gtk_widget_show (grid);
 	gtk_container_add (GTK_CONTAINER (window), grid);
 
-	nautilus_window_initialize_menus (window);
+	menubar = nautilus_window_initialize_menus (window);
 	nautilus_window_initialize_actions (window);
 
+	if (menubar) {
+		gtk_container_add (GTK_CONTAINER (grid), menubar);
+		gtk_widget_show_all (menubar);
+	}
+
 	/* Register to menu provider extension signal managing menu updates */
 	g_signal_connect_object (nautilus_signaller_get_current (), "popup-menu-changed",
 			 G_CALLBACK (nautilus_window_load_extension_menus), window, G_CONNECT_SWAPPED);
